### PR TITLE
fix(training): account for DSA in model FLOPs

### PR DESCRIPTION
**Pull request by yaoyu-33**
## Bug

`GLM5Bridge` selects `experimental_attention_variant="dsa"`, but the model-FLOPs calculator previously had no plain-DSA branch. It therefore fell through to dense MLA: the core used average context `S/2`, ignored DSA top-k and GLM-5.2 IndexShare cadence, and omitted the lightning indexer.

The fix is limited to the Bridge-side training FLOPs numerator. It does not change the B200 denominator, report projected MFU as measured performance, modify model-card metrics, or touch Megatron-Core.

## Sources and accounting

Primary architecture sources:

- [DeepSeek-V3.2 report](https://arxiv.org/abs/2512.02556), especially the lightning-indexer score and selected-attention equations, and [DeepSeek's official DSA reference release](https://github.com/deepseek-ai/DeepSeek-V3.2-Exp)
- [GLM-5 report](https://arxiv.org/abs/2602.15763)
- [GLM-5.2 architecture note](https://huggingface.co/blog/zai-org/glm-52-blog), which specifies one indexer per four backbone layers and first-step index computation for MTP
- The immutable [GLM-5.2 config at revision `4d67f66`](https://huggingface.co/zai-org/GLM-5.2/blob/4d67f66cc64d3219133b767c253b2ad1425c6c88/config.json)

Maintained training implementations compared:

- Current Megatron-Core [DSA](https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/transformer/experimental_attention_variant/dsa.py), [absorbed MLA](https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/transformer/experimental_attention_variant/absorbed_mla.py), and [indexer op](https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/transformer/experimental_attention_variant/ops/indexer.py)
- NVIDIA NeMo AutoModel's maintained [GLM DSA training layers](https://github.com/NVIDIA-NeMo/Automodel/blob/main/nemo_automodel/components/models/glm_moe_dsa/layers.py)

For fixed sequence length `S` and top-k `K`, the exact average causal contexts are:

```text
C_dense = (S + 1) / 2
C_sparse = (S + 1) / 2                              if S <= K
           K - K(K - 1) / (2S)                      otherwise
```

For `L` DSA layers and `L_i` full indexer layers, the added branch accounts for:

- absorbed sparse MLA QK/AV: `6 L C_sparse N_h ((R_kv + D_rope) + R_kv)`
- indexer projections: `4 L_i (Q_in H_i D_i + H D_i + H H_i)`
- dense index scores and weighted head reduction: `6 L_i C_dense H_i (D_i + 1)`
- sparse teacher QK for the indexer loss: `2 L_i C_sparse N_h (R_kv + D_rope)`

These are per-token terms. The factors include 2 FLOPs/FMA. Main attention uses forward+dgrad+wgrad. Megatron-Core detaches the indexer's `x` and query residual, so trained indexer projections use forward+wgrad (factor 4), while their dense score operands use forward+dgrad+wgrad (factor 6). The teacher Q/K are detached, so teacher QK is forward-only (factor 2). When the auxiliary loss is disabled, the indexer runs under `no_grad`, and its projection/score factors become 2.

GLM-5.2 has 78 decoder layers plus MTP1. Its `(freq=4, offset=3)` schedule yields 21 full backbone indexers; MTP numbering restarts and MTP1 computes one more, for `L_i=22`. Every DSA layer still pays sparse MLA work. The pre-existing DSV4 compressed-attention expression is not reused.

The calculator excludes ReLU, normalization, and top-k comparisons from the matmul-oriented numerator. Fixed-length results are exact under this convention. Packed metadata provides only `sum(s)` and `sum(s^2)`; when mixed packed subsequences cross top-k, the branch uses the existing token-weighted effective length, an explicit approximation. MTP retains the calculator's existing full-sequence convention.

## Red to green

Before the implementation, the new focused class demonstrated the fallthrough:

```text
uv run python -m pytest tests/unit_tests/training/utils/test_flop_utils.py::TestDynamicSparseAttentionFlops -q
5 failed
```

The failures showed the old dense-MLA value (`28680` instead of the DSA toy result `30028`), zero response to top-k/cadence/loss configuration, and the old MTP delta (`32328` instead of `33676`). After the implementation, the same five tests pass.

The tests independently derive exact toy totals and cover sequence/top-k scaling, IndexShare frequency and offset, detached-input backward multipliers, and MTP1. Existing exact MLA and DSV4 tests are retained as non-DSA regressions.

## GLM-5.2 before/after

Training TFLOP per sequence, batch size 1, MTP1. "Old" reproduces the former generic dense-MLA fallthrough; the component columns independently sum to the new DSA attention subtotal.

| Sequence | Old model total | New model total | Old dense MLA attention | MLA projections | Sparse QK | Sparse AV | Index projections | Dense index scores | Detached teacher | New DSA attention |
|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|
| 2,048 | 542.710 | 585.624 | 192.769 | 160.196 | 36.663 | 32.589 | 1.689 | 1.143 | 3.403 | 235.683 |
| 131,072 | 166,067.902 | 46,660.202 | 143,671.660 | 10,252.521 | 4,653.890 | 4,136.791 | 108.096 | 4,680.657 | 432.007 | 24,263.961 |
| 200,000 | 360,458.698 | 74,991.848 | 326,284.745 | 15,644.105 | 7,120.534 | 6,329.363 | 164.941 | 10,897.974 | 660.978 | 40,817.895 |

Historical verification-card TFLOPS remain unchanged because the original run logs were not recomputed; those values were emitted by the former formula.

## Validation

```text
uv run python -m pytest tests/unit_tests/training/utils/test_flop_utils.py -q
125 passed

uv run pre-commit run --all-files
all hooks passed

git diff --check
passed
```

### CODE REVIEW

**Review by bvolpato — APPROVED** [review]
LGTM. I traced IndexShare cadence, MTP layer numbering, and indexer grad/no-grad paths against current MCore. Full FLOP utility file passed: 125 tests.

Pinned GLM-5.2 recipe reproduces PR table at 2,048, 131,072, and 200,000 tokens. I also exercised GB200 LoRA config with TP=8, ETP=1, EP=64; calculation stays finite and picks up expected TP-driven vocabulary padding. Mixed packed sequences crossing top-k remain approximate, as documented.

**Review by bvolpato — APPROVED** [review]
LGTM. Full FLOP utility file passed: 125 tests. Pinned GLM-5.2 recipe reproduced the PR table exactly at 2,048, 131,072, and 200,000 tokens. TP=8, ETP=1, EP=64 also produced a finite result with expected TP-driven vocabulary padding.

This was analytical recipe/config evaluation, not a GB200 training run. PR changes FLOP accounting only, so there is no CUDA kernel path to exercise. Mixed packed sequences crossing top-k remain approximate, as documented.

File: src/megatron/bridge/training/utils/flop_utils.py [hunk 1]
```diff
@@ -881,6 +881,138 @@ def transformer_flops():
...
+                indexer_projection_multiplier = 4 if trains_indexer else 2
+                indexer_projection_term = indexer_projection_multiplier * num_indexer_layers * indexer_projection_size
+
+                # Equation 1 computes H_i dense q_i.k_i dot products and a weighted head
+                # reduction. When trained, both score operands need gradients (3x forward).
+                # ReLU, normalization, and top-k comparisons are not floating-point matmuls and
+                # are intentionally outside this model-FLOPs numerator.
+                index_score_size = dense_causal_context * idx_n_heads * (idx_head_dim + 1)
+                index_score_multiplier = 6 if trains_indexer else 2
+                index_score_term = index_score_multiplier * num_indexer_layers * index_score_size
```
Comment (HollowMan6) [inline src/megatron/bridge/training/utils/flop_utils.py]:
```suggestion
                # Equation 1 computes H_i dense q_i.k_i dot products and a weighted head
                # reduction. Top-k selection needs every causal score, so the forward is
                # always dense. The backward only covers score entries the KL loss touches:
                # every causal pair for the dense loss, but only the selected top-k pairs for
                # the sparse loss (no gradient flows through the discrete top-k selection;
                # MCore's indexer_backward_wrapper consumes the selected payload only).
                # ReLU, normalization, and top-k comparisons are not floating-point matmuls and
                # are intentionally outside this model-FLOPs numerator.
                use_sparse_indexer_loss = getattr(cfg.model, "dsa_indexer_use_sparse_loss", False)
                index_score_unit = idx_n_heads * (idx_head_dim + 1)
                index_score_term = 2 * num_indexer_layers * dense_causal_context * index_score_unit
                if trains_indexer:
                    score_grad_context = sparse_causal_context if use_sparse_indexer_loss else dense_causal_context
                    index_score_term += 4 * num_indexer_layers * score_grad_context * index_score_unit

```

File: src/megatron/bridge/training/utils/flop_utils.py [hunk 2]
```diff
@@ -881,6 +881,138 @@ def transformer_flops():
...
+                # GLM recipes enable MCore's sparse indexer KL loss. Its attention target uses
+                # detached main-model Q/K, hence forward-only QK work. Dense-loss configurations
+                # use the full causal context instead of the selected context.
+                indexer_teacher_term = 0
+                if trains_indexer:
+                    teacher_context = (
+                        sparse_causal_context
+                        if getattr(cfg.model, "dsa_indexer_use_sparse_loss", False)
+                        else dense_causal_context
+                    )
```
Comment (HollowMan6) [inline src/megatron/bridge/training/utils/flop_utils.py]:
```suggestion
                    teacher_context = sparse_causal_context if use_sparse_indexer_loss else dense_causal_context
```

File: tests/unit_tests/training/utils/test_flop_utils.py [hunk 3]
```diff
@@ -2122,6 +2126,113 @@ def test_mla_seq_length_quadratic_growth(self):
...
+        """A one-layer toy covers absorbed sparse MLA and every lightning-indexer matmul."""
+        # At S=4 and top-k=2, the causal selected counts are [1, 2, 2, 2],
+        # while the dense indexer sees [1, 2, 3, 4]. The independently reduced
+        # per-token terms are:
+        #   MLA projections: 906; sparse QK+AV: 147
+        #   indexer projections (forward+wgrad): 192
+        #   dense index scores (forward+dgrad+wgrad): 90
+        #   sparse detached teacher QK (forward only): 28
+        #   vocabulary projection: 6144
+        expected = 4 * (906 + 147 + 192 + 90 + 28 + 6144)
```
Comment (HollowMan6) [inline tests/unit_tests/training/utils/test_flop_utils.py]:
```suggestion
        #   index scores (dense forward 30 + sparse-loss backward 42): 72
        #   sparse detached teacher QK (forward only): 28
        #   vocabulary projection: 6144
        expected = 4 * (906 + 147 + 192 + 72 + 28 + 6144)
```

File: tests/unit_tests/training/utils/test_flop_utils.py [hunk 4]
```diff
@@ -2122,6 +2126,113 @@ def test_mla_seq_length_quadratic_growth(self):
...
+        short = num_floating_point_operations(self._dsa_config(), batch_size=1)
+        long = num_floating_point_operations(self._dsa_config(seq_length=8), batch_size=1)
+        wider_topk = num_floating_point_operations(self._dsa_config(dsa_indexer_topk=4), batch_size=1)
+
+        # S=8, k=2: average sparse context is 15/8 and dense causal context is 9/2.
+        assert long == 60_732
+        assert long > 2 * short
+        # S=4, k=4 raises average sparse context from 7/4 to 5/2. Only sparse
+        # QK/AV and the sparse teacher target change; top-k comparisons are not FLOPs.
+        assert wider_topk - short == 300
```
Comment (HollowMan6) [inline tests/unit_tests/training/utils/test_flop_utils.py]:
```suggestion
        assert long == 60_228
        assert long > 2 * short
        # S=4, k=4 raises average sparse context from 7/4 to 5/2. Sparse QK/AV,
        # the sparse-loss score backward, and the sparse teacher target change;
        # top-k comparisons are not FLOPs.
        assert wider_topk - short == 372
```

File: tests/unit_tests/training/utils/test_flop_utils.py [hunk 5]
```diff
@@ -2122,6 +2126,113 @@ def test_mla_seq_length_quadratic_growth(self):
...
+                dsa_indexer_topk_freq=4,
+                dsa_indexer_skip_topk_offset=1,
+            ),
+            batch_size=1,
+        )
+
+        # Full layers are [1..6], [1,2,3], and [1,5], respectively. Each full
+        # layer contributes (192 + 90 + 28) FLOPs per token of indexer work.
+        assert no_sharing - offset_three == 3 * 4 * 310
+        assert offset_three - offset_one == 4 * 310
```
Comment (HollowMan6) [inline tests/unit_tests/training/utils/test_flop_utils.py]:
```suggestion
        # layer contributes (192 + 72 + 28) FLOPs per token of indexer work.
        assert no_sharing - offset_three == 3 * 4 * 292
        assert offset_three - offset_one == 4 * 292
```

File: tests/unit_tests/training/utils/test_flop_utils.py [hunk 6]
```diff
@@ -2122,6 +2126,113 @@ def test_mla_seq_length_quadratic_growth(self):
...
+
+    def test_dsa_detached_indexer_projection_backward_multiplier(self):
+        """Indexer loss adds wgrad, not dgrad, for projections fed by detached inputs."""
+        with_loss = num_floating_point_operations(self._dsa_config(), batch_size=1)
+        without_loss = num_floating_point_operations(self._dsa_config(dsa_indexer_loss_coeff=0.0), batch_size=1)
+
+        # Enabling sparse indexer loss adds one projection wgrad (96/token),
+        # two score-operand gradients (60/token), and teacher QK (28/token).
+        assert with_loss - without_loss == 4 * (96 + 60 + 28)
```
Comment (HollowMan6) [inline tests/unit_tests/training/utils/test_flop_utils.py]:
```suggestion
        # score gradients over the selected top-k context only (42/token),
        # and teacher QK (28/token).
        assert with_loss - without_loss == 4 * (96 + 42 + 28)
```

File: tests/unit_tests/training/utils/test_flop_utils.py [hunk 7]
```diff
@@ -2122,6 +2126,113 @@ def test_mla_seq_length_quadratic_growth(self):
...
+
+    def test_dsa_mtp_layer_has_independent_sparse_attention_and_indexer(self):
+        """MTP1 adds one full DSA layer because MCore restarts MTP layer numbering at one."""
+        decoder_only = num_floating_point_operations(self._dsa_config(), batch_size=1)
+        with_mtp = num_floating_point_operations(self._dsa_config(mtp_num_layers=1), batch_size=1)
+
+        # Added per-token work: DSA layer 1363 + MTP norms/eh-proj 912 + logits 6144.
+        assert with_mtp - decoder_only == 4 * (1363 + 912 + 6144)
```
Comment (HollowMan6) [inline tests/unit_tests/training/utils/test_flop_utils.py]:
```suggestion
        # Added per-token work: DSA layer 1345 + MTP norms/eh-proj 912 + logits 6144.
        assert with_mtp - decoder_only == 4 * (1345 + 912 + 6144)
```

**Review by HollowMan6 — COMMENTED** [review]
The forward is rightly dense (top-k needs every score), but the backward should follow the loss mode: when `dsa_indexer_use_sparse_loss=True`, the KL loss only touches the selected top-k scores, but currently it charges the score backward over dense_causal_context in both loss modes, which is an over estimation.

Otherwise LGTM.
